### PR TITLE
Merge AccessibilityBridge and AccessibilityBridgeDelegate

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR merges `AccessibilityBridgeDelegate` into `AccessibilityBridge`, affecting `common`, `mac`, and `windows`. As a result, accessibility bridges and node delegates no longer call the engine for the bridge or the view, but get them as constructor arguments. `AccessibilityBridge` classes still refer to the engine, but only to dispatch events to embedder APIs.

Part of [the multiwindow project](https://github.com/flutter/flutter/issues/30701) ([design doc for macOS](https://docs.google.com/document/d/1U_ciMTHJgDrnw8kIo9sGKhiKFOycWRmz1yVVI_osCXg/edit)): This reduces the usage of the following single-view APIs:
- `mac/FlutterEngine#accessibilityBridge`
- `mac/FlutterEngine#viewController`
- `windows/FlutterWindowsEngine::accessibilityBridge`
- `windows/FlutterWindowsEngine::viewController`

TODO:
- [x] Isolate FlutterPlatformNodeDelegateMac from the engine
- [x] Apply the same change to the Windows embedding
- [x] Documentations change
- [ ] Update [design doc](https://docs.google.com/document/d/1qtt4v-DrS5gL8bcMlMqwq6xmc_-yRk2uYomfvGDyQTw/edit?resourcekey=0-IFvZRP2YpFrECaxxsQTyvQ#heading=h.4qsx9pjitde1) (internal doc)

### Why do we choose this method?

In the current HEAD:
* The bridge [must be initialized with a delegate](https://github.com/flutter/engine/blob/c16e2c08724c18d9275fcbcbc937af44e59e58a7/shell/platform/common/accessibility_bridge.h#L111) and owns it.
* The delegate is initialized with the engine.
* The delegate also [needs to use the bridge](https://github.com/flutter/engine/blob/c16e2c08724c18d9275fcbcbc937af44e59e58a7/shell/platform/darwin/macos/framework/Source/AccessibilityBridgeMacDelegate.mm#L51), however, not by directly referring to the bridge, but by calling `flutter_engine_.accessibilityBridge`.

The ultimate goal of the change is get rid of the usage of `FlutterEngine#accessibilityBridge` and `FlutterEngine#viewController`. Instead, the delegate will be provided a bridge and a view controller at initialization. In this way, the engine can initialize multiple sets of bridge-and-delegates, each for a window.

However, if we still keep the bridge and the delegate separated, the bridge must be _created_ with a delegate, and a delegate must be _created_ with a bridge. To resolve this circular dependency, we might either assign the delegate with the bridge later, or assign the bridge with the delegate later. Neither way is error-prone enough though: we can not guarantee at compile time that both classes have all dependents ready when other methods are called.

But since both the bridge and the delegate correspond to _one_ window, there's no reason to separate them. If we merge the two classes, then their communication becomes intra-class.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by dkwingsmt** [comment]
@chunhtai This PR takes an alternative approach from https://github.com/flutter/engine/pull/36573 by merge `AccessibilityBridgeDelegate` into `AccessibilityBridge` so that there are no circular referencing problem. 

I haven't implemented everything yet (see PR body for TODOs), however the current state should be a sufficient POC that the "merging" strategy works. Let me know if you think it's worth moving forward.

**Review by chunhtai — COMMENTED** [review]
I have a bit concern about using inheritence vs composition which may add more complexity if embedding starts overriding existing default method. but I think this is probably a better way to approach the dependency cycle problem. Should probably loop in @cbracken who wrote the code to wire up bridge in windows embedding.

**Comment by dkwingsmt** [comment]
Can you illustrate what you mean by 
> using inheritence vs composition which may add more complexity if embedding starts overriding existing default method

I don't think I'm using composition anywhere. The `FlutterEngine`'s new method is just for testing and is also overridden by inheritence.

**Comment by chunhtai** [comment]
> Can you illustrate what you mean by
> 
> > using inheritence vs composition which may add more complexity if embedding starts overriding existing default method

I meant it used to be composition (the bridge compose bridge delegate to call native API), but it becomes inheritance now. which may be a bit hard to maintain if embedding starts overriding method like AddFlutterSemanticsNodeUpdate or CommitUpdates to inject their own call loop. It is not a strong opinion to not to merge delegate and bridge though.

**Comment by cbracken** [comment]
@dkwingsmt can you elaborate a bit on the circular reference issue you mention in the comments above?

**Comment by dkwingsmt** [comment]
@cbracken Sure. The circular dependency is found in https://github.com/flutter/engine/pull/36573.

* Currently:
  * The bridge [owns a delegate](https://github.com/flutter/engine/pull/36573/files#diff-9f979d53c3e186c5c7ab9e9f71924b50535d029f0f994ceafc21fb1f88011c01R417).
  * The delegate is initialized with the engine.
  * The delegate also calls several methods of the bridge, however, not by directly referring to the bridge, but by calling `flutter_engine_.accessibilityBridge` ([here](https://github.com/flutter/engine/pull/36573/files#diff-7808506b685ffe5ea62d799f7b4614dfda0af1fee5f7edb5523e8c2b13b36fd2L51)).
* The ultimate goal of the change is get rid of the usage of `FlutterEngine#accessibilityBridge` and `FlutterEngine#viewController`. Instead, the delegate uses [a provided bridge](https://github.com/flutter/engine/pull/36573/files#diff-7808506b685ffe5ea62d799f7b4614dfda0af1fee5f7edb5523e8c2b13b36fd2R26) and [a provided view controller](https://github.com/flutter/engine/pull/36573/files#diff-7808506b685ffe5ea62d799f7b4614dfda0af1fee5f7edb5523e8c2b13b36fd2R33).
* But in this way, the bridge must be _created_ with a delegate, and a delegate must be _created_ with a bridge. This is the circular dependency.
* In the previous PR, I tried either assigning the delegate with the bridge later, or assigning the bridge with the delegate later. But neither is so error-prone.
* I discovered that, since both the bridge and the delegate correspond to _one_ window, there is no reason to separate them. Therefore in this PR, I propose to merge the two classes, resolving the circular dependency.

**Comment by dkwingsmt** [comment]
@cbracken @chunhtai This PR should be ready to review. Although, I might need some help on the documentations, since it feels like I should merge the two classes' the documentation but I don't know enough to do so.

File: shell/platform/common/accessibility_bridge.h [hunk 1]
```diff
@@ -168,12 +120,14 @@ class AccessibilityBridge
...
-  const std::vector<ui::AXEventGenerator::TargetedEvent> GetPendingEvents();
+  const std::vector<ui::AXEventGenerator::TargetedEvent> GetPendingEvents()
+      const;
 
+ protected:
   //------------------------------------------------------------------------------
   /// @brief      Update the AccessibilityBridgeDelegate stored in the
   ///             accessibility bridge to a new one.
-  void UpdateDelegate(std::unique_ptr<AccessibilityBridgeDelegate> delegate);
+  void Reset();
```
Comment (loic-sharma) [inline shell/platform/common/accessibility_bridge.h:130]: The comment above needs to be updated

File: shell/platform/common/accessibility_bridge_unittests.cc [hunk 2]
```diff
@@ -37,9 +37,8 @@ FlutterSemanticsNode CreateSemanticsNode(
 }
 
 TEST(AccessibilityBridgeTest, basicTest) {
-  std::shared_ptr<AccessibilityBridge> bridge =
-      std::make_shared<AccessibilityBridge>(
-          std::make_unique<TestAccessibilityBridgeDelegate>());
+  std::shared_ptr<TestAccessibilityBridge> bridge =
+      std::make_shared<TestAccessibilityBridge>();
```
Comment (loic-sharma) [inline shell/platform/common/accessibility_bridge_unittests.cc:41]: Just curious, do you know why these are `std::shared_ptr`s? It looks like they could be `std::unique_ptr`s instead

File: shell/platform/common/accessibility_bridge_unittests.cc [hunk 3]
```diff
@@ -108,20 +104,18 @@ TEST(AccessibilityBridgeTest, canFireChildrenChangedCorrectly) {
...
               Contains(ui::AXEventGenerator::Event::CHILDREN_CHANGED));
   EXPECT_THAT(actual_event,
               Contains(ui::AXEventGenerator::Event::SUBTREE_CREATED));
 }
 
 TEST(AccessibilityBridgeTest, canUpdateDelegate) {
-  std::shared_ptr<AccessibilityBridge> bridge =
-      std::make_shared<AccessibilityBridge>(
-          std::make_unique<TestAccessibilityBridgeDelegate>());
+  std::shared_ptr<TestAccessibilityBridge> bridge =
```
Comment (loic-sharma) [inline shell/platform/common/accessibility_bridge_unittests.cc:117]: This test name should be updated

File: shell/platform/common/test_accessibility_bridge.h [hunk 4]
```diff
@@ -22,6 +21,8 @@ class TestAccessibilityBridgeDelegate
   std::shared_ptr<FlutterPlatformNodeDelegate>
   CreateFlutterPlatformNodeDelegate() override;
 
+  void Update() { Reset(); }
```
Comment (loic-sharma) [inline shell/platform/common/test_accessibility_bridge.h:24]: Maybe call this `ResetBridge` to be closer to the parent's name?

File: shell/platform/windows/accessibility_bridge_windows_unittests.cc [hunk 5]
```diff
@@ -142,6 +157,14 @@ ui::AXNode* AXNodeFromID(std::shared_ptr<AccessibilityBridge> bridge,
   return node_delegate ? node_delegate->GetAXNode() : nullptr;
 }
 
+std::shared_ptr<AccessibilityBridgeWindowsSpy> getAccessibilityBridgeSpy(
```
Comment (loic-sharma) [inline shell/platform/windows/accessibility_bridge_windows_unittests.cc:160]:
```suggestion
std::shared_ptr<AccessibilityBridgeWindowsSpy> GetAccessibilityBridgeSpy(
```

**Review by loic-sharma — APPROVED** [review]
LGTM

File: shell/platform/windows/accessibility_bridge_windows.h [hunk 6]
```diff
@@ -2,59 +2,59 @@
...
-// The Win32 implementation of AccessibilityBridgeDelegate.
-//
-// Handles requests from the accessibility bridge to interact with Windows
-// accessibility APIs. This includes routing accessibility events fired from
-// the framework to Windows, routing native Windows accessibility events to the
-// framework, and creating Windows-specific FlutterPlatformNodeDelegate objects
-// for each node in the semantics tree.
-class AccessibilityBridgeDelegateWindows
-    : public AccessibilityBridge::AccessibilityBridgeDelegate {
+// The Win32 implementation of AccessibilityBridge.
```
Comment (loic-sharma) [inline shell/platform/windows/accessibility_bridge_windows.h:18]: Consider keeping the rest of the documentation from before:

```suggestion
// This interacts with Windows accessibility APIs, which includes routing
// accessibility events fired from the framework to Windows, routing native
// Windows accessibility events to the framework, and creating Windows-specific
// FlutterPlatformNodeDelegate objects for each node in the semantics tree.
```

Comment (dkwingsmt) [reply shell/platform/windows/accessibility_bridge_windows.h:18]: Thank you. I'm copying (and adapting) this paragraph to the mac bridge as well.

File: shell/platform/common/accessibility_bridge_unittests.cc (hunk 2 above)
Comment (dkwingsmt) [reply shell/platform/common/accessibility_bridge_unittests.cc:41]: Done.

File: shell/platform/common/accessibility_bridge.h [hunk 7]
```diff
@@ -31,85 +31,17 @@ namespace flutter {
 /// FlutterPlatformNodeDelegate to wrap each AXNode in order to provide
 /// an accessibility tree in the native format.
 ///
-/// This class takes in a AccessibilityBridgeDelegate instance and is in charge
-/// of its lifecycle. The delegate are used to handle the accessibility events
-/// and actions.
-///
-/// To use this class, you must provide your own implementation of
-/// FlutterPlatformNodeDelegate and AccessibilityBridgeDelegate.
+/// To use this class, you must implement this class and provide your own
```
Comment (chunhtai) [inline shell/platform/common/accessibility_bridge.h:34]:
```suggestion
/// To use this class, one must subclass this class and provide their own
```

File: shell/platform/common/accessibility_bridge.h [hunk 8]
```diff
@@ -168,12 +100,39 @@ class AccessibilityBridge
...
   ///             case one may decide to handle an event differently based on
   ///             all pending events.
-  const std::vector<ui::AXEventGenerator::TargetedEvent> GetPendingEvents();
+  const std::vector<ui::AXEventGenerator::TargetedEvent> GetPendingEvents()
+      const;
+
+ protected:
+  //---------------------------------------------------------------------------
+  /// @brief      Handle accessibility events generated due to accessibility
+  ///             tree changes. These events are generated in accessibility
```
Comment (chunhtai) [inline shell/platform/common/accessibility_bridge.h:112]:
```suggestion
  ///             tree changes. These events
```

File: shell/platform/common/accessibility_bridge.h [hunk 9]
```diff
@@ -168,12 +100,39 @@ class AccessibilityBridge
...
+  ///             bridge and needed to be sent to native accessibility system.
```
Comment (chunhtai) [inline shell/platform/common/accessibility_bridge.h:113]:
```suggestion
  ///             needed to be sent to native accessibility system.
```

File: shell/platform/common/accessibility_bridge.h (hunk 7 above)
Comment (chunhtai) [reply shell/platform/common/accessibility_bridge.h:34]: should it also mention the onaccessibilityevent?

**Review by chunhtai — APPROVED** [review]
LGTM, thanks for cleaning this up

File: shell/platform/common/accessibility_bridge.h (hunk 7 above)
Comment (dkwingsmt) [reply shell/platform/common/accessibility_bridge.h:34]: I'm not sure, can you help me write it?

**Pull request merged**